keps/sig-network/0031-service-topology.md

```diff
@@ -0,0 +1,259 @@
+* Performance benefit (node local/rack local) is lower latency/higher bandwidth.
+
+### Background
+
+It's a pain point for multi-zone clusters deployment since cross-zone network traffic being charged, while in-zone is not. In addition, cross-node traffic may carry sensitive metadata from other nodes. Therefore, users always prefer the service backends that close to them, e.g. same zone, rack and host etc. for security, performance and cost concerns.
+
+Kubernetes scheduler can constraining a pod to only be able to run on particular nodes/zones. However, Kubernetes service proxy just randomly picks an available backend for service routing and this one can be very far from the user, so we need a topology-aware service routing solution in Kubernetes. Basically, to find the nearest service backend. In other words, allowing people to configure if ALWAY reach a to local service backend. In this way, they can reduce network latency, improve security, save money and so on. However, because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this feature in Kubernetes, this document would like to introduce a more generic way to support arbitrary service topology.
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+Locality is an "user-defined" thing. When we set topology key "hostname" for service, we expect node carries different node labels on the key "hostname".
+
+Users can control the level of topology. For example, if someone run logging agent as a daemonset, he can set the "hard" topology requirement for same-host. If "hard" is not met, then just return "service not available". 
+
+And if someone set a "soft" topology requirement for same-host, say he "preferred" same-host endpoints and can accept other hosts when for some reasons local service's backend is not available on some host.
+
+If multiple endpoints satisfy the "hard" or "soft" topology requirement, we will randomly pick one by default. Routing decision is expected to be implemented in L3/4 VIP level such as kube proxy.
+
+
+## Implementation history
+
+### API changes
+
+The user need a way to declare which service is "local service" and what is the "topology key" of "local service".
+
+This will be accomplished through a new type object `ServicePolicy`.
```

Comment:
e.g. Just add a single field to `ServiceSpec`:

```
// Preference-order list of topology keys.  How this list is interpreted is implementation-defined.
//  E.g. a valid implementation might be to treat this field as a hard constraint: if backends exist for
// index [0], they will always be chosen; only if no backends exist for index [0] will backends
// for index [1] be considered.  Another valid implementation might be to monitor load and
// "spill over" into less-preferred topologies dynamically.
// 
// If this field is specified and all indices have no backends, the service has no backends, and
// connections will fail.
Topologies []string
```

(the field name can be debated)

We could argue that we need the "fall back to global" mode option (hard/soft), but even that could be satisfied by simply labeling all nodes or maybe by allowing "" as a special case meaning "match all nodes".

Less API is more good.  How little can we get away with?

Reply:
Slice of topology key would be simpler in API and we can get ride of new resource. But seems it's not straightforward for me to express HARD requirement. Can you give an example?

Reply:
Hard:

```

topologyKey: ["kubernetes.io/hostename"]

```

Soft:

```

topologyKey: ["kubernetes.io/hostname", ""]

```

This seems pretty natural. But my point above is asking which is more common, or more *expected*, hard or soft? If we want it to be soft unless the user is explicit about wanting it be a hard requirement (which is "safer" in a sense), then we need to treat the first one as soft and do something to flag that it's hard.

Reply:
It could be worthwhile to solicit more input on user expectations for this, great point.

My assumption (with little data) is that "hard" is the more common case.  Most use cases have been "give me my node-local agent".

It's not that I am making the last value special in its last-ness.  I am saying that (yeah, it's a little ugly) "" is an implicit label that every node satisfies.  E.g. if I wrote:

```
topologyKey: ["","kubernetes.io/hostname"]
```

Callers would never ever evaluate the "hostname" key because the "" key would match any node, and should never be empty (except in a 0-node cluster).

So yeah, it's "special" but by it's meaning, not it's position.

Reply:
I understand that it's not position, and I think it's easy to figure out what it means when you see it. I just think it may be non-obvious that in order to have it match "any" you need an empty string in there. If the name were `topologyConstraints` or `topologiesAllowed` or something similar (it's really just labels not inherently "topology") then my concern goes away. The `topologyKey` name comes from the podspec, which has additional context beyond just that name.